### PR TITLE
fix: nats consumer configurations tuning and code refactoring

### PR DESCRIPTION
**Pull request by gireesh-devtron**

### CODE REVIEW

File: pubsub-lib/NatsClient.go [hunk 1]
```diff
@@ -18,36 +18,59 @@
...
-	//stream wise
-	NatsStreamConfig string `env:"NATS_STREAM_CONFIG" envDefault:"{\"max_age\":86400000000000}"`
+	// NatsMsgBufferSize is the number of messages that will be buffered in memory (channel size)
+	// it is recommended to set this value equal to NatsMsgProcessingBatchSize as we want to process maximum messages in the buffer in one go.
+	// Note: if NatsMsgBufferSize is less than NatsMsgProcessingBatchSize
+	// then the wait time for the unprocessed messages in the buffer will be high.(total process time = life-time in buffer + processing time)
+	// NatsMsgBufferSize can be configured independently of NatsMsgProcessingBatchSize if needed by setting its value to positive value in env.
+	// if NatsMsgBufferSize set to a non-positive value then it will take the value of NatsMsgProcessingBatchSize.
+	// Note: always get this value by calling GetNatsMsgBufferSize method
+	NatsMsgBufferSize    int `env:"NATS_MSG_BUFFER_SIZE" envDefault:"1"`
```
Comment (gireesh-devtron) [inline pubsub-lib/NatsClient.go:47]: set to -1 as default

File: pubsub-lib/PubSubClientService.go [hunk 2]
```diff
@@ -77,7 +79,11 @@ func (impl PubSubClientServiceImpl) Publish(topic string, msg string) error {
 	return nil
 }
 
-func (impl PubSubClientServiceImpl) Subscribe(topic string, callback func(msg *model.PubSubMsg)) error {
+// Subscribe method is used to subscribe to the given topic(+required),
+// this creates blocking process to continuously fetch messages from nats server published on this topic.
+// invokes callback(+required) func for each message received.
+// validations(+optional) methods were called before passing the message to the callback func.
+func (impl PubSubClientServiceImpl) Subscribe(topic string, callback func(msg *model.PubSubMsg), loggerFunc func(msg *model.PubSubMsg), validations ...ValidateMsg) error {
```
Comment (gireesh-devtron) [inline pubsub-lib/PubSubClientService.go:86]: add comments about logger func

File: pubsub-lib/PubSubClientService.go [hunk 3]
```diff
@@ -9,14 +9,16 @@ import (
 	"github.com/devtron-labs/common-lib/utils"
 	"github.com/nats-io/nats.go"
 	"go.uber.org/zap"
+	"k8s.io/utils/pointer"
 	"runtime/debug"
 	"sync"
 	"time"
 )
 
+type ValidateMsg func(msgId *string) bool
```
Comment (gireesh-devtron) [inline pubsub-lib/PubSubClientService.go:18]: remove pointer to msgId

File: pubsub-lib/PubSubClientService.go [hunk 4]
```diff
@@ -190,14 +177,35 @@ func (impl PubSubClientServiceImpl) publishPanicError(msg *nats.Msg, panicErr er
...
 // TryCatchCallBack is a fail-safe method to use callback function
-func (impl PubSubClientServiceImpl) TryCatchCallBack(msg *nats.Msg, callback func(msg *model.PubSubMsg)) {
-	subMsg := &model.PubSubMsg{Data: string(msg.Data)}
+func (impl PubSubClientServiceImpl) TryCatchCallBack(msg *nats.Msg, callback func(msg *model.PubSubMsg), loggerFunc func(msg *model.PubSubMsg), validations ...ValidateMsg) {
+	var msgDeliveryCount uint64 = 0
+	if metadata, err := msg.Metadata(); err == nil {
+		msgDeliveryCount = metadata.NumDelivered
+	}
+	natsMsgId := msg.Header.Get(model.NatsMsgId)
+	var natsMsgIdPtr *string = nil
```
Comment (gireesh-devtron) [inline pubsub-lib/PubSubClientService.go:186]: remove this

File: pubsub-lib/PubSubClientService.go [hunk 5]
```diff
@@ -190,14 +177,35 @@ func (impl PubSubClientServiceImpl) publishPanicError(msg *nats.Msg, panicErr er
...
+			return
+		}
+	}
 	defer func() {
 		// Acknowledge the message delivery
 		err := msg.Ack()
 		if err != nil {
 			impl.Logger.Errorw("nats: unable to acknowledge the message", "subject", msg.Subject, "msg", string(msg.Data))
 		}
+		metrics.NatsEventDeliveryCount.WithLabelValues(msg.Subject, natsMsgId).Observe(float64(msgDeliveryCount))
```
Comment (gireesh-devtron) [inline pubsub-lib/PubSubClientService.go:208]: add this metric only for redelivery

File: pubsub-lib/PubSubClientService.go [hunk 6]
```diff
@@ -235,3 +243,37 @@ func (impl PubSubClientServiceImpl) getStreamConfig(streamName string) *nats.Str
...
+	if err != nil {
+		impl.Logger.Errorw("unable to retrieve consumer info from NATS-server", "stream", streamName, "consumer", consumerName, "err", err)
+		return
+	}
+
+	existingConfig := info.Config
+	updatesDetected := false
+
+	// Currently only checking for AckWait,MaxAckPending but can be done for other editable properties as well
+	if overrideConfig.AckWaitInSecs > 0 && existingConfig.AckWait != time.Duration(overrideConfig.AckWaitInSecs)*time.Second {
```
Comment (gireesh-devtron) [inline pubsub-lib/PubSubClientService.go:262]: refactor this

**Review by kripanshdevtron — APPROVED** [review]

**Pull request merged**